### PR TITLE
Add onOpenPrepackagedDatabase callback

### PR DESCRIPTION
**Pull request by mzgreen**
This patch adds a new builder methods that accept new PrepackagedCallback for creating a database.
The new PrepackagedCallback contains a single method onOpenPrepackagedDatabase(SupportSQLiteDatabase) which is called right after creating a db from Asset/File/InputStream.

onOpenPrepackagedDatabase(SupportSQLiteDatabase) callback is transactional which means that it'll be called only once after creating a db from Asset/File/InputStream. If an exception is thrown inside onOpenPrepackagedDatabase callback then the next time this database is opened, it should be created from Asset/File/InputStream again and onOpenPrepackagedDatabase should also be called again.

Test: Added a unit tests to PrepackageTest.
Fixes: b/148934423

### CODE REVIEW

File: room/integration-tests/testapp/src/androidTest/java/androidx/room/integration/testapp/test/DatabaseCallbackTest.java
Comment (tikurahul) [inline room/integration-tests/testapp/src/androidTest/java/androidx/room/integration/testapp/test/DatabaseCallbackTest.java]: Might want to do this in an `@After` block so tests can be idempotent.

File: room/runtime/api/api_lint.ignore
Comment (tikurahul) [inline room/runtime/api/api_lint.ignore]: You should not be adding these to the ignore list. Use `@Suppress` instead. We added some notes in `Contributing.md` for this.

**Review by tikurahul — COMMENTED** [review]
These are just initial thoughts on the PR. @danysantiago will look at the API in more detail.

File: room/runtime/api/restricted_current.txt
Comment (yigit) [inline room/runtime/api/restricted_current.txt:8]: we cannot remove these constructors for backwards compatibility (because generated code w/ older version of room might be calling it).
so need to add the parameter to the end and chain.
@danysantiago to double check in case this does not get called by generated code. (though seems like we've been chaining these: https://cs.android.com/androidx/platform/frameworks/support/+/androidx-master-dev:room/runtime/src/main/java/androidx/room/DatabaseConfiguration.java?q=DatabaseConfiguration)

File: room/runtime/src/main/java/androidx/room/SQLiteCopyOpenHelper.java
Comment (yigit) [inline room/runtime/src/main/java/androidx/room/SQLiteCopyOpenHelper.java]: can we extract this into another method to avoid code duplication?

File: room/runtime/api/api_lint.ignore
Comment (mzgreen) [reply room/runtime/api/api_lint.ignore]: I tried using annotations but `./gradlew updateApi` task was still failing on metalava lint. Adding those manually fixed the issue. But I'll double check this.

File: room/runtime/src/main/java/androidx/room/DatabaseConfiguration.java
Comment (danysantiago) [inline room/runtime/src/main/java/androidx/room/DatabaseConfiguration.java]: To be backwards compatible with old generated code, even though this API is marked with `@RestrictTo` we need to copy this constructor and add the new parameter and then update all other constructors to call the newest one with `null`. Here is an example: https://github.com/androidx/androidx/commit/1263185ef8c50d4059758e382610345a24f82c53#diff-a135d07f04ca20de890a27db682fe8e2

File: room/runtime/src/main/java/androidx/room/DatabaseConfiguration.java
Comment (danysantiago) [inline room/runtime/src/main/java/androidx/room/DatabaseConfiguration.java]: I'm thinking it might be best / simpler if instead of a list we only take one callback. The reason is we only allow one of the configurations `createFromFile` / `createFromAsset` / `createFromStream`, and since the callback is passed on those builder methods it seems weird that calling it again to change the configuration (to like update the file name) you end up passing the same callback and it'll get invoked twice.

File: room/integration-tests/testapp/src/androidTest/java/androidx/room/integration/testapp/test/DatabaseCallbackTest.java
Comment (danysantiago) [inline room/integration-tests/testapp/src/androidTest/java/androidx/room/integration/testapp/test/DatabaseCallbackTest.java]: Can you plz move these test to https://github.com/androidx/androidx/blob/androidx-master-dev/room/integration-tests/testapp/src/androidTest/java/androidx/room/integration/testapp/test/PrepackageTest.java? These tests are for the old callbacks and since we are scoping the new one I think it might be best to have it with the rest of prepackage DB tests.

File: room/integration-tests/testapp/src/androidTest/java/androidx/room/integration/testapp/test/DatabaseCallbackTest.java
Comment (danysantiago) [inline room/integration-tests/testapp/src/androidTest/java/androidx/room/integration/testapp/test/DatabaseCallbackTest.java]: If we move the tests to `PrepackageTest` we won't need to copy this.

File: room/runtime/api/api_lint.ignore
Comment (danysantiago) [reply room/runtime/api/api_lint.ignore]: Try `@SupressLint`, take a look at `createFromInputStream` which added the same suppression.

File: room/runtime/src/main/java/androidx/room/SQLiteCopyOpenHelper.java
Comment (danysantiago) [inline room/runtime/src/main/java/androidx/room/SQLiteCopyOpenHelper.java]: Lets not open the DB with the 'expected final version' we need to open the pre-package DB with the same version it already has, to avoid going through the migration path which will end up updating the version and when Room opens the DB again it will think its already updated. To find out the version of the DB without opening use https://github.com/androidx/androidx/blob/androidx-master-dev/room/runtime/src/main/java/androidx/room/util/DBUtil.java#L139

File: room/runtime/src/main/java/androidx/room/SQLiteCopyOpenHelper.java
Comment (mzgreen) [reply room/runtime/src/main/java/androidx/room/SQLiteCopyOpenHelper.java]: I wanted to extract this logic into a separate method and call it in `SQLiteCopyOpenHelper.getWritableDatabase()` and `SQLiteCopyOpenHelper.getReadableDatabase()` but when I temporarily open the db using `FrameworkSqliteOpenHelper` then I need to call `getReadableDatabase()` or `getWritableDatabase()` and I wasn't sure if I can just use one of those for both calls.
For example - can I open the db using framework helper, call `getWritableDatabase()` and pass it to the `onPrepackagedCallback` inside `SQLiteCopyOpenHelper.getReadableDatabase()`? Or I have to call `getReadableDatabase()` in this case?

File: room/runtime/src/main/java/androidx/room/SQLiteCopyOpenHelper.java
Comment (danysantiago) [inline room/runtime/src/main/java/androidx/room/SQLiteCopyOpenHelper.java]: I don't think we should catch all SQLiteException and ignore them.

File: room/runtime/src/main/java/androidx/room/SQLiteCopyOpenHelper.java
Comment (danysantiago) [inline room/runtime/src/main/java/androidx/room/SQLiteCopyOpenHelper.java]: `callOnOpenPrepackagedDatabaseCallbacks` should be inside a try-finally where in the finally block we close the DB.

File: room/runtime/src/main/java/androidx/room/SQLiteCopyOpenHelper.java
Comment (danysantiago) [reply room/runtime/src/main/java/androidx/room/SQLiteCopyOpenHelper.java]: You can still extract the logic into a function that receives a boolean parameter `writable` to know which of the two methods to call.

File: room/runtime/src/main/java/androidx/room/RoomDatabase.java
Comment (danysantiago) [inline room/runtime/src/main/java/androidx/room/RoomDatabase.java:1218]: We should add a bit more docs here: "This callback will be invoked after the pre-package DB is copied but before Room had a chance to open it and therefore before the RoomDatabase.Callback methods are invoked. This callback can be useful for updating the pre-package DB schema to satisfy Room's schema validation."

**Review by danysantiago — CHANGES_REQUESTED** [review]
This is great, thank you for making this pull request!

File: room/runtime/src/main/java/androidx/room/SQLiteCopyOpenHelper.java
Comment (mzgreen) [reply room/runtime/src/main/java/androidx/room/SQLiteCopyOpenHelper.java]: I appreciate those explanations. It helps a lot. Thanks!

File: room/runtime/src/main/java/androidx/room/SQLiteCopyOpenHelper.java
Comment (mzgreen) [reply room/runtime/src/main/java/androidx/room/SQLiteCopyOpenHelper.java]: What do you think we should do with the exceptions? Would it be enough to print some log here?

Comment (danysantiago) [reply room/runtime/src/main/java/androidx/room/SQLiteCopyOpenHelper.java]: We should let it propagate up and crash the app (or let the user of Room catch it), its the same as exceptions in `onOpen`, they would crash the app.

File: room/runtime/api/api_lint.ignore
Comment (mzgreen) [reply room/runtime/api/api_lint.ignore]: I think the problem is that for example `createFromFile` needs 2 suppressions: `BuilderSetStyle` and `StreamFiles`. If I add only one then it's picked up by the tooling but it still complains about second issue. When I do `@SuppressLint("BuilderSetStyle, StreamFiles")` then it doesn't work at all, same effect as not adding any suppression. I think this should work since the value of `SuppressLint` is an array of strings.

Comment (dlam) [reply room/runtime/api/api_lint.ignore]: I'm not sure it's possible to pass an array of string via varargs to the annotation? Have you tried:
`@SuppressLint({"BuilderSetStyle", "StreamFiles"})`? This is how we do it elsewhere in the codebase.

Comment (dlam) [reply room/runtime/api/api_lint.ignore]: Oh, I just noticed you are passing a single string that's comma separated :smiley: 

Definitely use `@SuppressLint({"BuilderSetStyle", "StreamFiles"})` instead.

Comment (mzgreen) [reply room/runtime/api/api_lint.ignore]: Ah Kotlin habits.. You're right! It worked, thanks :)

File: room/integration-tests/testapp/src/androidTest/java/androidx/room/integration/testapp/test/DatabaseCallbackTest.java
Comment (mzgreen) [reply room/integration-tests/testapp/src/androidTest/java/androidx/room/integration/testapp/test/DatabaseCallbackTest.java]: I agree but I wanted to keep new tests consistent with the existing ones so I kept it this way

File: room/runtime/src/main/java/androidx/room/SQLiteCopyOpenHelper.java
Comment (mzgreen) [reply room/runtime/src/main/java/androidx/room/SQLiteCopyOpenHelper.java]: I made the changes but I'm still not sure if it's ok. There are a couple of things that may throw:
1. Reading the db version
2. Calling `getReadable/WritableDatabase()`

Take a look please and let me know if you'd like to change anything

**Comment by dlam** [comment]
It looks like you may want to rebase your changes here or some kind of weird merge issue? The diff is really large now and includes a bunch of changes outside of room.

**Comment by mzgreen** [comment]
I did a rebase and now Github shows more changes than it should 😕 
Seems like similar issue: https://stackoverflow.com/questions/16306012/github-pull-request-showing-commits-that-are-already-in-target-branch
There is no other branch that I could use to try the suggested workaround.

Not sure what can I do about it. If I try to open a new PR then Github correctly shows my changes but I don't know how to refresh this one.

**Comment by danysantiago** [comment]
You can try cherry-picking only the relevant commits on top a new checkout of androidx force-pushing to the branch of this PR. This also seems relevant: https://stackoverflow.com/questions/16306012/github-pull-request-showing-commits-that-are-already-in-target-branch/45286781#45286781, but beware I'm not a GitHub guru. xD

**Comment by dlam** [comment]
If you want to give me force push access on your fork as well, I can try to fix it :smiley:

**Comment by dlam** [comment]
Should be fixed now, let me know if I missed anything - the issue was mixing merge + rebase flows :slightly_smiling_face: 

i.e., tried to rebase a merge commit on top of origin/androidx-master-dev

**Comment by mzgreen** [comment]
I see, thanks @dlam !

File: room/integration-tests/testapp/src/androidTest/java/androidx/room/integration/testapp/test/PrepackageTest.java
Comment (danysantiago) [inline room/integration-tests/testapp/src/androidTest/java/androidx/room/integration/testapp/test/PrepackageTest.java]: Lets make this an int and increment it every time the callback is called, that way we can assert it gets called exactly once.

File: room/runtime/src/main/java/androidx/room/SQLiteCopyOpenHelper.java
Comment (danysantiago) [inline room/runtime/src/main/java/androidx/room/SQLiteCopyOpenHelper.java]: * missing space after `if`
* Also invert the condition so that we can return early and save indentation:
```
if (mDatabaseConfiguration.prepackagedCallback == null) {
  return
}
```

File: room/runtime/src/main/java/androidx/room/SQLiteCopyOpenHelper.java
Comment (danysantiago) [inline room/runtime/src/main/java/androidx/room/SQLiteCopyOpenHelper.java]: Maybe call method `dispatchOnOpenPrepackagedDatabase`

File: room/runtime/src/main/java/androidx/room/SQLiteCopyOpenHelper.java
Comment (danysantiago) [inline room/runtime/src/main/java/androidx/room/SQLiteCopyOpenHelper.java]: `SQLiteException` is an unchecked exception, lets just not catch it and let it propagate up, since wrapping it in a RuntimeException makes it harder for the user to handle it.

File: room/runtime/src/main/java/androidx/room/SQLiteCopyOpenHelper.java
Comment (danysantiago) [inline room/runtime/src/main/java/androidx/room/SQLiteCopyOpenHelper.java]: nit: missing space in if: `if (version == null) {`

File: room/runtime/src/main/java/androidx/room/SQLiteCopyOpenHelper.java
Comment (danysantiago) [inline room/runtime/src/main/java/androidx/room/SQLiteCopyOpenHelper.java]: Let's change the message to 'Malformed database file, unable to read version'. It would also be better to use the `DBUtil` method directly here and catch the IOException so that you can propagate it with the new RuntimeException, it provides a bit more info if something goes wrong.

File: room/runtime/src/main/java/androidx/room/SQLiteCopyOpenHelper.java
Comment (danysantiago) [inline room/runtime/src/main/java/androidx/room/SQLiteCopyOpenHelper.java]: Not sure this method is necessary, in `callOnOpenPrepackagedDatabaseCallbacks` where it is now newly used we should use the `DBUtil` version so that we can add the `IOException` as the cause to the `RuntimeException`.

File: room/runtime/src/main/java/androidx/room/RoomDatabase.java
Comment (danysantiago) [inline room/runtime/src/main/java/androidx/room/RoomDatabase.java]: nit: I think `// pre-dates rule.` fits in the previous line.

File: room/integration-tests/testapp/src/androidTest/java/androidx/room/integration/testapp/test/PrepackageTest.java
Comment (danysantiago) [inline room/integration-tests/testapp/src/androidTest/java/androidx/room/integration/testapp/test/PrepackageTest.java]: If we update `TestPrepackagedCallback` to use an int for counting everytime the callback is called then we don't need `callback2` making this slightly simpler.

**Review by danysantiago — CHANGES_REQUESTED** [review]
This is looking great, just a few more comments.

File: room/runtime/src/main/java/androidx/room/SQLiteCopyOpenHelper.java
Comment (mzgreen) [reply room/runtime/src/main/java/androidx/room/SQLiteCopyOpenHelper.java]: Is there an auto formatting task that I could run or maybe code formatting config file that I could import in AS? The default one seems to be using different rules.

Comment (danysantiago) [reply room/runtime/src/main/java/androidx/room/SQLiteCopyOpenHelper.java]: I'm not sure, the Studio that is started with `studiow` should have the style configuration. I know we also have some tasks to run `ktlint` but I run them when uploading to Gerrit using the `repo` tool which is different from this Github flow. @dlam might know more about this. It's possible this might not be setup yet, sorry.

**Review by danysantiago — APPROVED** [review]
Looks good to me, thanks!

**Review by tikurahul — APPROVED** [review]

**Comment by mzgreen** [comment]
Thanks!

Not sure if it matters but the description at the top contains an error. It says that tests were added to `DatabaseCallbackTest` but I later moved them to `PrepackageTest` file and forgot to update Github description. This description was copied to Gerrit, I tried to edit it there but I can't (probably missing permission).

**Comment by dlam** [comment]
Does your gerrit account use the same e-mail as your git commit author? Copybara not updating the gerrit CL description sounds like a bug, thanks for letting us know.

**Review by dlam — APPROVED** [review]

**Comment by dlam** [comment]
Looks like Copybara was just waiting for review approvals before syncing.

**Comment by dlam** [comment]
The CL in gerrit seems to have been updated with the new description now.

**Comment by mzgreen** [comment]
I think Copybara works fine. I was trying to update the description on Gerrit before and it didn't work. Now after your message I realized that Copybara will copy it from Github if I change it here so I updated it and it worked. I understand the process now.

**Pull request closed without merging**
